plugins/dell-dock/fu-dell-dock-plugin.c

```diff
@@ -99,80 +119,297 @@ fu_dell_dock_plugin_probe(FuPlugin *plugin, FuDevice *proxy, GError **error)
+	}
+
+	return TRUE;
+}
+
 static FuDevice *
-fu_dell_dock_plugin_get_ec(GPtrArray *devices)
+fu_dell_dock_plugin_get_ec_device(FuPlugin *plugin)
 {
-	FuDevice *ec_parent = NULL;
+	GPtrArray *devices = fu_plugin_get_devices(plugin);
+
 	for (gint i = devices->len - 1; i >= 0; i--) {
 		FuDevice *dev = g_ptr_array_index(devices, i);
-		FuDevice *parent;
+		if (FU_IS_DELL_DOCK_EC_V2(dev))
+			return dev;
 		if (FU_IS_DELL_DOCK_EC(dev))
 			return dev;
-		parent = fu_device_get_parent(dev);
-		if (parent != NULL && FU_IS_DELL_DOCK_EC(parent))
-			ec_parent = parent;
 	}
-
-	return ec_parent;
+	return NULL;
 }
 
 static gboolean
-fu_dell_dock_plugin_backend_device_added(FuPlugin *plugin,
-					 FuDevice *device,
-					 FuProgress *progress,
-					 GError **error)
+fu_dell_dock_plugin_device_add(FuPlugin *plugin, FuDevice *device, gboolean newdev, GError **error)
 {
-	g_autoptr(FuDeviceLocker) locker = NULL;
-	g_autoptr(FuDellDockHub) hub = NULL;
-	const gchar *hub_cache_key = "hub-usb3-gen1";
-	GPtrArray *devices;
 	FuDevice *ec_device;
-	FuDevice *hub_dev;
 	guint8 dock_type;
+	guint16 vid, pid;
+	g_autofree const gchar *instance_id = NULL;
 
-	/* not interesting */
-	if (!FU_IS_USB_DEVICE(device))
+	/* VID and PID */
+	vid = fu_usb_device_get_vid(FU_USB_DEVICE(device));
+	pid = fu_usb_device_get_pid(FU_USB_DEVICE(device));
+
+	/* cache current device until EC dock type is available */
+	ec_device = fu_dell_dock_plugin_get_ec_device(plugin);
+	if (ec_device == NULL) {
+		const gchar *key;
+		key = g_strdup_printf("USB\\VID_%04X&PID_%04X", vid, pid);
+		fu_plugin_cache_add(plugin, key, device);
 		return TRUE;
+	}
 
-	hub = fu_dell_dock_hub_new(FU_USB_DEVICE(device));
-	locker = fu_device_locker_new(FU_DEVICE(hub), error);
-	if (locker == NULL)
+	/* determine dock type said by ec */
+	dock_type = fu_dell_dock_plugin_get_dock_type(ec_device, error);
+	if (dock_type == DOCK_BASE_TYPE_UNKNOWN) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_READ,
+			    "can't read base dock type from EC");
 		return FALSE;
-
-	/* probe extend devices under Usb3.1 Gen 2 Hub */
-	if (fu_device_has_private_flag(FU_DEVICE(hub), FU_DELL_DOCK_HUB_FLAG_HAS_BRIDGE)) {
-		if (!fu_dell_dock_plugin_probe(plugin, FU_DEVICE(hub), error))
-			return FALSE;
 	}
 
-	/* process hub devices if ec device is added */
-	devices = fu_plugin_get_devices(plugin);
-	ec_device = fu_dell_dock_plugin_get_ec(devices);
-	if (ec_device == NULL) {
-		fu_plugin_cache_add(plugin, hub_cache_key, FU_DEVICE(hub));
-		return TRUE;
+	/* USB HUB Devices */
+	instance_id = fu_dell_dock_get_instance_id(&dock_type, dock_hwdev_info_hub, &vid, &pid);
+	if (instance_id != NULL) {
+		g_autofree const gchar *instance_guid = NULL;
+		g_autoptr(FuDellDockHub) hub_device = NULL;
+
+		if (newdev) {
+			hub_device = fu_dell_dock_hub_new(FU_USB_DEVICE(device));
+			if (!fu_device_open(FU_DEVICE(hub_device), error))
+				return FALSE;
+		} else {
+			hub_device = FU_DELL_DOCK_HUB(device);
```

Comment:
I've just realized that this assignment (or even with g_steal_pointer) will fall into a situation that a ghost device in the plugin.  I don't have better way to do this but currently it is https://github.com/fwupd/fwupd/pull/7369/files#diff-cf3e2d3d7e960a425b0276b4438c42ef41d1d841fddafa0d256331654f303844R202.